File: app/backend/models/Exam.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.models;

import java.util.*;
import java.util.stream.Collectors;
import javax.annotation.Nonnull;
import javax.persistence.*;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.ebean.annotation.EnumValue;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.joda.time.DateTime;
import org.springframework.beans.BeanUtils;

import backend.models.api.AttachmentContainer;
import backend.models.base.OwnedModel;
import backend.models.questions.Question;
import backend.models.sections.ExamSection;
import backend.models.sections.ExamSectionQuestion;
import backend.models.sections.ExamSectionQuestionOption;
import backend.util.AppUtil;
import backend.util.datetime.DateTimeAdapter;


@Entity
public class Exam extends OwnedModel implements Comparable<Exam>, AttachmentContainer {

    public enum State {
        @EnumValue("1") DRAFT,
        @EnumValue("2") SAVED,
        @EnumValue("3") PUBLISHED,       // EXAM PUBLISHED, VISIBLE TO STUDENTS AND READY FOR TAKING
        @EnumValue("4") STUDENT_STARTED, // EXAM STARTED BY STUDENT
        @EnumValue("5") REVIEW,          // EXAM RETURNED BY STUDENT AND READY FOR REVIEW
        @EnumValue("6") REVIEW_STARTED,  // REVIEW STARTED BY TEACHERS
        @EnumValue("7") GRADED,          // GRADE GIVEN
        @EnumValue("13") PRE_PUBLISHED,  // COLLABORATIVE EXAM READY FOR TEACHERS FOR EDITING
        /* FINAL STATES */
        @EnumValue("8") GRADED_LOGGED,   // EXAM PROCESSED AND READY FOR REGISTRATION
        @EnumValue("9") ARCHIVED,        // EXAM ARCHIVED FOR CERTAIN PERIOD AFTER WHICH IT GETS DELETED
        @EnumValue("10") ABORTED,        // EXAM ABORTED BY STUDENT WHILST TAKING
        @EnumValue("11") DELETED,        // EXAM MARKED AS DELETED AND HIDDEN FROM END USERS
        @EnumValue("12") REJECTED        // EXAM NOT QUALIFIED FOR REGISTRATION
    }

    private boolean anonymous;

    private String name;

    @ManyToOne
    private Course course;

    @ManyToOne
    private ExamType examType;

    @ManyToMany
    @JoinTable(name = "exam_owner", joinColumns = @JoinColumn(name = "exam_id"), inverseJoinColumns = @JoinColumn(name = "user_id"))
    private Set<User> examOwners;

    // Instruction written by teacher, shown during exam
    @Column(columnDefinition = "TEXT")
    private String instruction;

    // Instruction written by teacher, shown for reservation purposes
    @Column(columnDefinition = "TEXT")
    private String enrollInstruction;

    private boolean shared;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "exam")
    @JsonManagedReference
    private Set<ExamSection> examSections;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "exam")
    @JsonManagedReference
    private Set<ExaminationDate> examinationDates;

    @ManyToOne(cascade = CascadeType.PERSIST)
    protected Exam parent;

    @OneToMany(mappedBy = "parent")
    @JsonBackReference
    protected List<Exam> children;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "exam")
    @JsonManagedReference
    private List<ExamEnrolment> examEnrolments;

    @OneToOne(mappedBy = "exam")
    @JsonManagedReference
    private ExamParticipation examParticipation;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "exam")
    @JsonManagedReference
    private Set<ExamInspection> examInspections;

    @OneToOne(mappedBy = "exam")
    private ExamRecord examRecord;

    @OneToOne(mappedBy = "exam", cascade = CascadeType.ALL)
    private AutoEvaluationConfig autoEvaluationConfig;

    @OneToOne(mappedBy = "exam")
    private LanguageInspection languageInspection;

    @Column(length = 32, unique = true)
    private String hash;

    // Exam valid/enrollable from
    @Temporal(TemporalType.TIMESTAMP)
    @JsonSerialize(using = DateTimeAdapter.class)
    private DateTime examActiveStartDate;

    // Exam valid/enrollable until
    @Temporal(TemporalType.TIMESTAMP)
    @JsonSerialize(using = DateTimeAdapter.class)
    private DateTime examActiveEndDate;

    // Exam duration (minutes)
    private Integer duration;

    @ManyToOne
    private GradeScale gradeScale;

    // Custom course credit - if teachers changes course credit
    private Double customCredit;

    // Exam language
    @ManyToMany
    private List<Language> examLanguages;

    // Exam answer language
    private String answerLanguage;

    private State state;

    @ManyToOne
    private Grade grade;

    // Ohjelmistot
    @ManyToMany(cascade = CascadeType.ALL)
    private List<Software> softwares;

    /*
     * this is the user who is marked as evaluator of the Exam
     * in WebOodi, or other system
     */
    @ManyToOne
    private User gradedByUser;

    @Temporal(TemporalType.TIMESTAMP)
    @JsonSerialize(using = DateTimeAdapter.class)
    private DateTime gradedTime;

    @OneToOne
    private Comment examFeedback;

    private String additionalInfo;

    // Number of times a student is allowed to take the exam before getting a grade
    private Integer trialCount;

    @ManyToOne
    private ExamType creditType;

    @ManyToOne
    private ExamExecutionType executionType;

    @OneToMany(mappedBy = "exam", cascade = CascadeType.ALL)
    private Set<InspectionComment> inspectionComments;

    // In UI, section has been expanded
    @Column(columnDefinition = "boolean default false")
    private boolean expanded;

    @OneToOne(cascade = CascadeType.ALL)
    private Attachment attachment;

    @Temporal(TemporalType.TIMESTAMP)
    private DateTime autoEvaluationNotified;

    private boolean gradeless;

    private Boolean subjectToLanguageInspection;

    // Optional internal reference to this exam
    private String internalRef;

    private String assessmentInfo;

    public User getGradedByUser() {
        return gradedByUser;
    }

    public void setGradedByUser(User gradedByUser) {
        this.gradedByUser = gradedByUser;
    }

    public Set<User> getExamOwners() {
        return examOwners;
    }

    public void setExamOwners(Set<User> examOwners) {
        this.examOwners = examOwners;
    }

    public Set<ExaminationDate> getExaminationDates() {
        return examinationDates;
    }

    public void setExaminationDates(Set<ExaminationDate> examinationDates) {
        this.examinationDates = examinationDates;
    }

    // Aggregate properties, required as fields by Ebean
    @Transient
    private Double totalScore;
    @Transient
    private Double maxScore;
    @Transient
    private int rejectedAnswerCount;
    @Transient
    private int approvedAnswerCount;

    @Transient
    private boolean cloned;
    @Transient
    private boolean external;
    @Transient
    private String externalRef;


    public Double getTotalScore() {
        return examSections.stream()
                .map(ExamSection::getTotalScore)
                .reduce(0.0, (sum, x) -> sum += x);
    }

    public Double getMaxScore() {
        return examSections.stream()
                .map(ExamSection::getMaxScore)
                .reduce(0.0, (sum, x) -> sum += x);
    }

    private int getApprovedAnswerCount() {
        return examSections.stream()
                .map(ExamSection::getApprovedCount)
                .reduce(0, (sum, x) -> sum += x);
    }

    private int getRejectedAnswerCount() {
        return examSections.stream()
                .map(ExamSection::getRejectedCount)
                .reduce(0, (sum, x) -> sum += x);
    }

    // These are dumb, required to be explicitly set by EBean
    public void setTotalScore() {
        totalScore = getTotalScore();
    }

    public void setMaxScore() {
        maxScore = getMaxScore();
    }

    public void setRejectedAnswerCount() {
        rejectedAnswerCount = getRejectedAnswerCount();
    }

    public void setApprovedAnswerCount() {
        approvedAnswerCount = getApprovedAnswerCount();
    }

    public boolean isCloned() {
        return cloned;
    }

    public void setCloned(boolean cloned) {
        this.cloned = cloned;
    }

    public boolean isExternal() {
        return external;
    }

    public void setExternal(boolean external) {
        this.external = external;
    }

    public String getExternalRef() {
        return externalRef;
    }

    public void setExternalRef(String externalRef) {
        this.externalRef = externalRef;
    }

    public DateTime getGradedTime() {
        return gradedTime;
    }

    public void setGradedTime(DateTime gradedTime) {
        this.gradedTime = gradedTime;
    }

    public boolean getExpanded() {
        return expanded;
    }

    public void setExpanded(boolean expanded) {
        this.expanded = expanded;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public ExamType getExamType() {
        return examType;
    }

    public void setExamType(ExamType examType) {
        this.examType = examType;
    }

    public Set<ExamSection> getExamSections() {
        return examSections;
    }

    public void setExamSections(Set<ExamSection> examSections) {
        this.examSections = examSections;
    }

    public String getInstruction() {
        return instruction;
    }

    public void setInstruction(String instruction) {
        this.instruction = instruction;
    }

    public boolean isShared() {
        return shared;
    }

    public void setShared(boolean shared) {
        this.shared = shared;
    }

    public Course getCourse() {
        return course;
    }

    public void setCourse(Course course) {
        this.course = course;
    }

    public String getHash() {
        return hash;
    }

    public void setHash(String hash) {
        this.hash = hash;
    }

    public Integer getDuration() {
        return duration;
    }

    public void setDuration(Integer duration) {
        this.duration = duration;
    }

    public GradeScale getGradeScale() {
        return gradeScale;
    }

    public void setGradeScale(GradeScale gradeScale) {
        this.gradeScale = gradeScale;
    }

    public Double getCustomCredit() {
        return customCredit;
    }

    public void setCustomCredit(Double customCredit) {
        this.customCredit = customCredit;
    }

    public List<Language> getExamLanguages() {
        return examLanguages;
    }

    public void setExamLanguages(List<Language> examLanguages) {
        this.examLanguages = examLanguages;
    }

    public String getAnswerLanguage() {
        return answerLanguage;
    }

    public void setAnswerLanguage(String answerLanguage) {
        this.answerLanguage = answerLanguage;
    }

    public String generateHash() {
        String attributes = name + state + new Random().nextDouble();
        hash = AppUtil.encodeMD5(attributes);
        return hash;
    }

    public String getEnrollInstruction() {
        return enrollInstruction;
    }

    public void setEnrollInstruction(String enrollInstruction) {
        this.enrollInstruction = enrollInstruction;
    }

    public Exam getParent() {
        return parent;
    }

    public void setParent(Exam parent) {
        this.parent = parent;
    }

    public Grade getGrade() {
        return grade;
    }

    public void setGrade(Grade grade) {
        this.grade = grade;
    }

    public State getState() {
        return state;
    }

    public void setState(State state) {
        this.state = state;
    }

    public Integer getTrialCount() {
        return trialCount;
    }

    public void setTrialCount(Integer trialCount) {
        this.trialCount = trialCount;
    }

    public Comment getExamFeedback() {
        return examFeedback;
    }

    public void setExamFeedback(Comment examFeedback) {
        this.examFeedback = examFeedback;
    }

    public String getAdditionalInfo() {
        return additionalInfo;
    }

    public void setAdditionalInfo(String additionalInfo) {
        this.additionalInfo = additionalInfo;
    }

    public ExamType getCreditType() {
        return creditType;
    }

    public void setCreditType(ExamType creditType) {
        this.creditType = creditType;
    }

    public List<Software> getSoftwareInfo() {
        return softwares;
    }

    public void setSoftwareInfo(List<Software> softwareInfo) {
        softwares = softwareInfo;
    }

    public List<ExamEnrolment> getExamEnrolments() {
        return examEnrolments;
    }

    public void setExamEnrolments(List<ExamEnrolment> examEnrolments) {
        this.examEnrolments = examEnrolments;
    }

    public ExamParticipation getExamParticipation() {
        return examParticipation;
    }

    public void setExamParticipation(ExamParticipation examParticipation) {
        this.examParticipation = examParticipation;
    }

    public List<Exam> getChildren() {
        return children;
    }

    public void setChildren(List<Exam> children) {
        this.children = children;
    }

    public Set<ExamInspection> getExamInspections() {
        return examInspections;
    }

    public void setExamInspections(Set<ExamInspection> examInspections) {
        this.examInspections = examInspections;
    }

    public LanguageInspection getLanguageInspection() {
        return languageInspection;
    }

    public ExamRecord getExamRecord() {
        return examRecord;
    }

    public void setLanguageInspection(LanguageInspection languageInspection) {
        this.languageInspection = languageInspection;
    }

    public String getAssessmentInfo() {
        return assessmentInfo;
    }

    public void setAssessmentInfo(String assessmentInfo) {
        this.assessmentInfo = assessmentInfo;
    }

    private Exam createCopy(User user, boolean produceStudentExam, boolean setParent) {
        Exam clone = new Exam();
        BeanUtils.copyProperties(this, clone, "id", "examSections", "examEnrolments", "examParticipation",
                "examInspections", "autoEvaluationConfig", "creator", "created", produceStudentExam ? "examOwners" : "none");
        if (setParent) {
            clone.setParent(this);
        }
        AppUtil.setCreator(clone, user);
        AppUtil.setModifier(clone, user);
        clone.generateHash();
        clone.save();

        if (autoEvaluationConfig != null) {
            AutoEvaluationConfig configClone = autoEvaluationConfig.copy();
            configClone.setExam(clone);
            configClone.save();
            clone.setAutoEvaluationConfig(configClone);
        }

        for (ExamInspection ei : examInspections) {
            ExamInspection inspection = new ExamInspection();
            BeanUtils.copyProperties(ei, inspection, "id", "exam");
            inspection.setExam(clone);
            inspection.save();
        }
        Set<ExamSection> sections = new TreeSet<>(examSections);
        for (ExamSection es : sections) {
            ExamSection esCopy = es.copy(clone, produceStudentExam, setParent);
            AppUtil.setCreator(esCopy, user);
            AppUtil.setModifier(esCopy, user);
            // Shuffle question options before saving
            for (ExamSectionQuestion esq : esCopy.getSectionQuestions()) {
                List<ExamSectionQuestionOption> shuffled = new ArrayList<>(esq.getOptions());
                Collections.shuffle(shuffled);
                esq.setOptions(new HashSet<>(shuffled));
            }
            esCopy.save();
            for (ExamSectionQuestion esq : esCopy.getSectionQuestions()) {
                if (produceStudentExam) {
                    Question questionCopy = esq.getQuestion();
                    AppUtil.setCreator(questionCopy, user);
                    AppUtil.setModifier(questionCopy, user);
                    questionCopy.update();
                }
                esq.save();
            }
            clone.getExamSections().add(esCopy);
        }
        if (attachment != null) {
            Attachment copy = new Attachment();
            BeanUtils.copyProperties(attachment, copy, "id");
            clone.setAttachment(copy);
        }
        return clone;
    }

    public Exam copyForStudent(User student, boolean isCollaborative) {
        return createCopy(student, true, !isCollaborative);
    }

    public Exam copy(User user) {
        return createCopy(user, false, true);
    }

    public DateTime getExamActiveStartDate() {
        return examActiveStartDate;
    }

    public void setExamActiveStartDate(DateTime examActiveStartDate) {
        this.examActiveStartDate = examActiveStartDate;
    }

    public DateTime getExamActiveEndDate() {
        return examActiveEndDate;
    }

    public void setExamActiveEndDate(DateTime examActiveEndDate) {
        this.examActiveEndDate = examActiveEndDate;
    }

    public void setAttachment(Attachment attachment) {
        this.attachment = attachment;
    }

    public Attachment getAttachment() {
        return attachment;
    }

    public ExamExecutionType getExecutionType() {
        return executionType;
    }

    public void setExecutionType(ExamExecutionType executionType) {
        this.executionType = executionType;
    }

    public AutoEvaluationConfig getAutoEvaluationConfig() {
        return autoEvaluationConfig;
    }

    public void setAutoEvaluationConfig(AutoEvaluationConfig autoEvaluationConfig) {
        this.autoEvaluationConfig = autoEvaluationConfig;
    }

    public Set<InspectionComment> getInspectionComments() {
        return inspectionComments;
    }

    public void setInspectionComments(Set<InspectionComment> inspectionComments) {
        this.inspectionComments = inspectionComments;
    }

    public DateTime getAutoEvaluationNotified() {
        return autoEvaluationNotified;
    }

    public void setAutoEvaluationNotified(DateTime autoEvaluationNotified) {
        this.autoEvaluationNotified = autoEvaluationNotified;
    }

    public boolean isGradeless() {
        return gradeless;
    }

    public void setGradeless(boolean gradeless) {
        this.gradeless = gradeless;
    }

    public Boolean getSubjectToLanguageInspection() {
        return subjectToLanguageInspection;
    }

    public void setSubjectToLanguageInspection(Boolean subjectToLanguageInspection) {
        this.subjectToLanguageInspection = subjectToLanguageInspection;
    }

    public String getInternalRef() {
        return internalRef;
    }

    public void setInternalRef(String internalRef) {
        this.internalRef = internalRef;
    }

    public boolean isAnonymous() {
        return anonymous;
    }

    public void setAnonymous(boolean anonymous) {
        this.anonymous = anonymous;
    }

    @Transient
    private boolean isCreatedBy(User user) {
        return creator != null && creator.equals(user);
    }

    @Transient
    private boolean isInspectedBy(User user, boolean applyToChildOnly) {
        Exam examToCheck = parent == null || applyToChildOnly ? this : parent;
        return examToCheck.examInspections.stream().anyMatch(ei -> ei.getUser().equals(user));
    }

    @Transient
    private boolean isOwnedBy(User user) {
        Exam examToCheck = parent == null ? this : parent;
        return examToCheck.examOwners.stream().anyMatch(owner -> owner.equals(user));
    }

    @Transient
    public boolean isOwnedOrCreatedBy(User user) {
        return isCreatedBy(user) || isOwnedBy(user);
    }

    @Transient
    public boolean isInspectedOrCreatedOrOwnedBy(User user) {
        return isInspectedBy(user, false) || isOwnedBy(user) || isCreatedBy(user);
    }

    @Transient
    public boolean isChildInspectedOrCreatedOrOwnedBy(User user) {
        return isInspectedBy(user, true) || isOwnedBy(user) || isCreatedBy(user);
    }

    @Transient
    public boolean isViewableForLanguageInspector(User user) {
        return executionType.getType().equals(ExamExecutionType.Type.MATURITY.toString()) &&
                user.hasPermission(Permission.Type.CAN_INSPECT_LANGUAGE) && languageInspection != null &&
                languageInspection.getAssignee() != null;
    }

    @Transient
    public boolean isPrivate() {
        return !executionType.getType().equals(ExamExecutionType.Type.PUBLIC.toString()) && !isPrintout();
    }

    @Transient
    public boolean isPrintout() {
        return executionType.getType().equals(ExamExecutionType.Type.PRINTOUT.toString());
    }

    @Transient
    public boolean hasState(State... states) {
        return Arrays.asList(states).contains(state);
    }

    @Transient
    public boolean hasOptionalSections() {
        return examSections.stream().anyMatch(ExamSection::isOptional);
    }

    @Transient
    public Set<ExamSection> getOptionalSections() {
        return examSections.stream().filter(ExamSection::isOptional).collect(Collectors.toSet());
    }

    @Transient
    public void setDerivedMaxScores() {
        examSections.stream()
                .flatMap(es -> es.getSectionQuestions().stream())
                .forEach(esq -> {
                    esq.setDerivedMaxScore();
                    esq.getOptions().forEach(o -> o.setScore(null));
                });
    }


    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Exam)) {
            return false;
        }
        Exam otherExam = (Exam) other;
        return new EqualsBuilder().append(id, otherExam.id).build();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id).build();
    }

    @Override
    public String toString() {
        return "Exam{" +
                "course=" + course +
                ", id='" + id + '\'' +
                ", name='" + name + '\'' +
                ", examType=" + examType +
                ", hash='" + hash + '\'' +
                ", state='" + state + '\'' +
                '}';
    }

    @Override
    public int compareTo(@Nonnull Exam other) {
        return created.compareTo(other.created);
    }


}

File: app/models/exam/Exam.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.exam

import com.fasterxml.jackson.annotation.{JsonBackReference, JsonManagedReference}
import features.exam.copy.ExamCopyContext
import jakarta.persistence.*
import models.assessment.*
import models.attachment.{Attachment, AttachmentContainer}
import models.base.OwnedModel
import models.enrolment.{ExamEnrolment, ExamParticipation, ExaminationEventConfiguration}
import models.facility.Software
import models.questions.QuestionType
import models.sections.{ExamSection, ExamSectionQuestion}
import models.user.{Language, PermissionType, User}
import org.apache.commons.codec.digest.DigestUtils
import services.datetime.JsonInstant

import java.text.{DecimalFormat, DecimalFormatSymbols}
import java.time.Instant
import java.util.*
import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*

@Entity
class Exam extends OwnedModel with Ordered[Exam] with AttachmentContainer:
  @OneToMany(cascade = Array(CascadeType.ALL), mappedBy = "exam")
  @JsonManagedReference
  var examSections: java.util.Set[ExamSection] = uninitialized

  @OneToMany(cascade = Array(CascadeType.ALL), mappedBy = "exam")
  @JsonManagedReference
  var examinationDates: java.util.Set[ExaminationDate] = uninitialized

  @OneToMany(cascade = Array(CascadeType.ALL), mappedBy = "exam")
  @JsonManagedReference
  var examinationEventConfigurations: java.util.Set[ExaminationEventConfiguration] = uninitialized

  @OneToMany(mappedBy = "parent")
  @JsonBackReference
  var children: java.util.List[Exam] = uninitialized

  @OneToMany(cascade = Array(CascadeType.ALL), mappedBy = "exam")
  @JsonManagedReference
  var examEnrolments: java.util.List[ExamEnrolment] = uninitialized

  @OneToMany(cascade = Array(CascadeType.ALL), mappedBy = "exam")
  @JsonManagedReference
  var examInspections: java.util.Set[ExamInspection] = uninitialized

  @OneToMany(mappedBy = "exam", cascade = Array(CascadeType.ALL))
  var inspectionComments: java.util.Set[InspectionComment] = uninitialized

  @ManyToMany
  @JoinTable(
    name = "exam_owner",
    joinColumns = Array(new JoinColumn(name = "exam_id")),
    inverseJoinColumns = Array(new JoinColumn(name = "user_id"))
  )
  var examOwners: java.util.Set[User] = uninitialized

  @OneToOne(mappedBy = "exam")
  @JsonManagedReference
  var examParticipation: ExamParticipation = uninitialized

  @OneToOne(mappedBy = "exam", cascade = Array(CascadeType.ALL))
  var autoEvaluationConfig: AutoEvaluationConfig = uninitialized

  @OneToOne(mappedBy = "exam", cascade = Array(CascadeType.ALL))
  var examFeedbackConfig: ExamFeedbackConfig = uninitialized

  @ManyToMany(cascade = Array(CascadeType.ALL)) var softwares: List[Software] = uninitialized
  @ManyToMany var examLanguages: java.util.List[Language]                     = uninitialized
  @ManyToOne var course: Course                                               = uninitialized
  @ManyToOne var examType: ExamType                                           = uninitialized
  @ManyToOne(cascade = Array(CascadeType.PERSIST)) var parent: Exam           = uninitialized
  @ManyToOne var gradeScale: GradeScale                                       = uninitialized
  @ManyToOne var grade: Grade                                                 = uninitialized
  @ManyToOne var creditType: ExamType                                         = uninitialized
  @ManyToOne var executionType: ExamExecutionType                             = uninitialized
  @ManyToOne var gradedByUser: User                                           = uninitialized
  @OneToOne var examFeedback: Comment                                         = uninitialized
  @OneToOne(cascade = Array(CascadeType.ALL)) var attachment: Attachment      = uninitialized
  @OneToOne(mappedBy = "exam") var examRecord: ExamRecord                     = uninitialized
  @OneToOne(mappedBy = "exam") var languageInspection: LanguageInspection     = uninitialized

  @JsonInstant
  @Column(name = "exam_active_start_date")
  var periodStart: Instant = uninitialized

  @JsonInstant
  @Column(name = "exam_active_end_date")
  var periodEnd: Instant = uninitialized

  @JsonInstant var gradedTime: Instant             = uninitialized
  @JsonInstant var autoEvaluationNotified: Instant = uninitialized

  var instruction: String                            = uninitialized
  var enrollInstruction: String                      = uninitialized
  var anonymous: Boolean                             = false
  var name: String                                   = uninitialized
  var shared: Boolean                                = false
  var hash: String                                   = uninitialized
  var duration: Integer                              = uninitialized
  var customCredit: java.lang.Double                 = uninitialized
  var answerLanguage: String                         = uninitialized
  var state: ExamState                               = uninitialized
  var implementation: ExamImplementation             = uninitialized
  var additionalInfo: String                         = uninitialized
  var trialCount: Integer                            = uninitialized
  var gradingType: GradeType                         = uninitialized
  var subjectToLanguageInspection: java.lang.Boolean = uninitialized
  var internalRef: String                            = uninitialized
  var assessmentInfo: String                         = uninitialized
  var organisations: String                          = uninitialized

  // Aggregate transient fields — explicitly set by Ebean
  @Transient var totalScore: Double       = uninitialized
  @Transient var maxScore: Double         = uninitialized
  @Transient var rejectedAnswerCount: Int = 0
  @Transient var approvedAnswerCount: Int = 0
  @Transient var cloned: Boolean          = false
  @Transient var external: Boolean        = false
  @Transient var externalRef: String      = uninitialized

  def getTotalScore: Double = math.max(toFixed(examSections.asScala.map(_.getTotalScore).sum), 0.0)
  def getMaxScore: Double   = toFixed(examSections.asScala.map(_.getMaxScore).sum)

  def setTotalScore(): Unit = totalScore = getTotalScore
  def setMaxScore(): Unit   = maxScore = getMaxScore
  def setRejectedAnswerCount(): Unit =
    rejectedAnswerCount = examSections.asScala.map(_.getRejectedCount).sum
  def setApprovedAnswerCount(): Unit =
    approvedAnswerCount = examSections.asScala.map(_.getApprovedCount).sum

  def generateHash(): Unit =
    hash = DigestUtils.md5Hex(name + state + new Random().nextDouble())

  def hasState(states: ExamState*): Boolean = states.contains(state)

  def isOwnedOrCreatedBy(user: User): Boolean = isCreatedBy(user) || isOwnedBy(user)

  def isInspectedOrCreatedOrOwnedBy(user: User): Boolean =
    isInspectedBy(user, applyToChildOnly = false) || isOwnedBy(user) || isCreatedBy(user)

  def isChildInspectedOrCreatedOrOwnedBy(user: User): Boolean =
    isInspectedBy(user, applyToChildOnly = true) || isOwnedBy(user) || isCreatedBy(user)

  def isViewableForLanguageInspector(user: User): Boolean =
    executionType != null &&
      executionType.`type` == ExamExecutionType.Type.MATURITY.toString &&
      user.hasPermission(PermissionType.CAN_INSPECT_LANGUAGE) &&
      languageInspection != null &&
      languageInspection.assignee != null

  def isPrivate: Boolean =
    executionType != null &&
      executionType.`type` != ExamExecutionType.Type.PUBLIC.toString &&
      !isPrintout

  def isPrintout: Boolean =
    executionType != null && executionType.`type` == ExamExecutionType.Type.PRINTOUT.toString

  def isUnsupervised: Boolean =
    executionType == null || executionType.`type` != ExamImplementation.AQUARIUM.toString

  def setDerivedMaxScores(): Unit =
    examSections.asScala
      .flatMap(_.sectionQuestions.asScala)
      .foreach { esq =>
        esq.setDerivedMaxScore()
        val qType = Option(esq.question).map(_.`type`)
        if qType.contains(QuestionType.ClaimChoiceQuestion) || qType.contains(
            QuestionType.WeightedMultipleChoiceQuestion
          )
        then
          esq.setDerivedMinScore()
        esq.options.asScala.foreach(_.score = 0)
      }

  def createCopy(context: ExamCopyContext): Exam =
    val clone = scalarCopy()
    if !context.shouldExcludeExamOwners then clone.examOwners = examOwners
    if context.shouldSetParent then clone.parent = this
    clone.setCreatorWithDate(context.getUser)
    clone.setModifierWithDate(context.getUser)
    clone.generateHash()
    clone.save()

    if autoEvaluationConfig != null then
      val configClone = autoEvaluationConfig.copy()
      configClone.exam = clone
      configClone.save()
      clone.autoEvaluationConfig = configClone

    for ei <- examInspections.asScala do
      val inspection = new ExamInspection
      inspection.user = ei.user
      inspection.assignedBy = ei.assignedBy
      inspection.comment = ei.comment
      inspection.ready = ei.ready
      inspection.exam = clone
      inspection.save()

    for es <- selectSectionsToCopy(context) do
      val esCopy = es.copy(clone, context)
      esCopy.setCreatorWithDate(context.getUser)
      esCopy.setModifierWithDate(context.getUser)
      if context.shouldShuffleOptions then
        esCopy.sectionQuestions.asScala.foreach(shuffleQuestionOptions)
      esCopy.save()
      if context.isStudentExam then updateQuestionMetadata(esCopy, context.getUser)
      clone.examSections.add(esCopy)

    if attachment != null then
      clone.attachment = attachment.copy()

    clone

  def scalarCopy(): Exam =
    val dest = new Exam
    dest.name = name
    dest.state = state
    dest.implementation = implementation
    dest.course = course
    dest.examType = examType
    dest.instruction = instruction
    dest.enrollInstruction = enrollInstruction
    dest.shared = shared
    dest.examinationDates = examinationDates
    dest.examRecord = examRecord
    dest.examFeedbackConfig = examFeedbackConfig
    dest.languageInspection = languageInspection
    dest.periodStart = periodStart
    dest.periodEnd = periodEnd
    dest.duration = duration
    dest.gradeScale = gradeScale
    dest.customCredit = customCredit
    dest.examLanguages = examLanguages
    dest.answerLanguage = answerLanguage
    dest.grade = grade
    dest.softwares = softwares
    dest.gradedByUser = gradedByUser
    dest.gradedTime = gradedTime
    dest.examFeedback = examFeedback
    dest.additionalInfo = additionalInfo
    dest.trialCount = trialCount
    dest.creditType = creditType
    dest.executionType = executionType
    dest.inspectionComments = inspectionComments
    dest.autoEvaluationNotified = autoEvaluationNotified
    dest.gradingType = gradingType
    dest.subjectToLanguageInspection = subjectToLanguageInspection
    dest.internalRef = internalRef
    dest.assessmentInfo = assessmentInfo
    dest.organisations = organisations
    dest.anonymous = anonymous
    dest.modifier = modifier
    dest.modified = modified
    dest

  private def toFixed(v: Double): Double = Exam.Df.format(v).toDouble

  private def selectSectionsToCopy(context: ExamCopyContext): Iterable[ExamSection] =
    val sections = examSections.asScala
    val filtered =
      if context.shouldIncludeOnlySelectedSections then
        sections.filter(es => !es.optional || context.getSelectedSections.contains(es.id))
      else sections
    filtered.toList.sorted

  private def updateQuestionMetadata(section: ExamSection, user: User): Unit =
    for esq <- section.sectionQuestions.asScala do
      val q = esq.question
      q.setCreatorWithDate(user)
      q.setModifierWithDate(user)
      q.update()
      esq.save()

  private def shuffleQuestionOptions(esq: ExamSectionQuestion): Unit =
    val shouldShuffle =
      esq.optionShufflingOn &&
        Option(esq.question).map(_.`type`).forall(_ != QuestionType.ClaimChoiceQuestion)
    if shouldShuffle then
      val shuffled = new ArrayList(esq.options)
      Collections.shuffle(shuffled)
      esq.options = shuffled

  private def isCreatedBy(user: User): Boolean = creator != null && creator == user

  private def isInspectedBy(user: User, applyToChildOnly: Boolean): Boolean =
    val examToCheck = if parent == null || applyToChildOnly then this else parent
    examToCheck.examInspections.asScala.exists(_.user == user)

  private def isOwnedBy(user: User): Boolean =
    val examToCheck = if parent == null then this else parent
    examToCheck.examOwners.asScala.exists(_ == user)

  override def compare(other: Exam): Int = created.compareTo(other.created)

  override def equals(o: Any): Boolean = o match
    case e: Exam => this.id == e.id
    case _       => false

  override def hashCode: Int = id.toInt

object Exam:
  type State          = ExamState
  type Implementation = ExamImplementation

  private val Df = new DecimalFormat("#.##", new DecimalFormatSymbols(Locale.US))
